File: source/equipment/systm.py
#
# MPowerTCX: Share Schwinn A.C. indoor cycle data with Strava, GoldenCheetah and other apps
# Copyright (C) 2017 James Roth
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import datetime

from . import bikes


class Systm(bikes.Bike):
    header = [
        'ticks',
        'videoTimestamp',
        'targetPower4dpType',
        'intervalId',
        'timeOfDayTimestamp',   # time_t + 3 digits
        'distanceMeters',
        'speedVirtualMps',
        'distanceVirtualMeters',
        'workoutPosition',
        'powerMatchOffset',
        'cadence',
        'speed',
        'heartRate',
        'power',
        'distanceSensorMeters',
        'trainerPower',
        'time',
        'targetPower',
        'targetHeartRateZone',
        'targetCadence',
        'targetRpe',
        'secondsSinceStart',
        'grade',
        'speedSensorMps',
        'leftPower',
        'rightPower'
    ]

    keys = dict(zip(header, list(range(len(header)))))

    def load(self, peek, reader):
        if peek == self.header:
            self._load(reader)
            return True

        return False

    def name(self):
        return "Wahoo SYSTM"

    def _get(self, row, field):
        idx = self.keys[field]
        value = row[idx]

        if value == "NaN":
            value = "0"

        return float(value)

    def _load(self, reader):
        last_time = 0.0
        start_stamp = None

        for row in reader:
            if len(row):
                if start_stamp is None:
                    start_stamp = int(self._get(row, 'timeOfDayTimestamp') / 1000)

                last_time = self._get(row, 'videoTimestamp')

                self.ride.addSample(
                    power=int(self._get(row, 'trainerPower')),
                    rpm=self._get(row, 'cadence'),
                    hr=self._get(row, 'heartRate'),
                    distance=self._get(row, 'distanceVirtualMeters')
                )

        if start_stamp is not None:
            when = datetime.datetime.fromtimestamp(start_stamp)
            self.ride.set_date_hint(when)

        self.ride.inferHeader(last_time)

File: crates/mpowertcx-core/src/equipment/systm.rs
use crate::equipment::{BikeParser, CsvRows};
use crate::ride::{python_float, Ride};

pub struct Systm;

impl Systm {
    pub fn new() -> Self {
        Self
    }

    fn header() -> &'static [&'static str] {
        &[
            "ticks", "videoTimestamp", "targetPower4dpType", "intervalId",
            "timeOfDayTimestamp", "distanceMeters", "speedVirtualMps",
            "distanceVirtualMeters", "workoutPosition", "powerMatchOffset",
            "cadence", "speed", "heartRate", "power", "distanceSensorMeters",
            "trainerPower", "time", "targetPower", "targetHeartRateZone",
            "targetCadence", "targetRpe", "secondsSinceStart", "grade",
            "speedSensorMps", "leftPower", "rightPower",
        ]
    }

    fn get(row: &[String], field: &str) -> f64 {
        let header = Self::header();
        if let Some(idx) = header.iter().position(|h| *h == field) {
            if let Some(val) = row.get(idx) {
                if val == "NaN" {
                    return 0.0;
                }
                return val.parse::<f64>().unwrap_or(0.0);
            }
        }
        0.0
    }

    fn load(rows: &mut CsvRows, ride: &mut Ride) {
        let mut last_time = 0.0f64;
        let mut start_stamp: Option<i64> = None;

        while let Some(row) = rows.next() {
            if row.is_empty() {
                continue;
            }
            if start_stamp.is_none() {
                start_stamp = Some(Self::get(row, "timeOfDayTimestamp") as i64 / 1000);
            }
            last_time = Self::get(row, "videoTimestamp");

            let power = Self::get(row, "trainerPower") as i64;
            let cadence = python_float(Self::get(row, "cadence"));
            let hr = python_float(Self::get(row, "heartRate"));
            let distance = python_float(Self::get(row, "distanceVirtualMeters"));

            ride.add_sample(power, cadence, hr, distance);
        }

        if let Some(stamp) = start_stamp {
            if let Some(dt) = chrono::DateTime::from_timestamp(stamp, 0) {
                ride.set_date_hint(dt.naive_local());
            }
        }

        ride.infer_header(last_time, python_float(last_time));
    }
}

impl BikeParser for Systm {
    fn try_load(&mut self, peek: &[String], rows: &mut CsvRows, ride: &mut Ride) -> bool {
        let header = Self::header();
        if peek.len() == header.len() && peek.iter().zip(header.iter()).all(|(a, b)| a == b) {
            Self::load(rows, ride);
            true
        } else {
            false
        }
    }

    fn name(&self) -> &str {
        "Wahoo SYSTM"
    }
}
